web3_proxy/src/admin_queries.rs

```diff
@@ -0,0 +1,125 @@
+    TypedHeader,
+};
+use axum::response::{IntoResponse, Response};
+use entities::{admin, user, user_tier};
+use ethers::prelude::Address;
+use hashbrown::HashMap;
+use http::StatusCode;
+use migration::sea_orm::{self, ActiveModelTrait, ColumnTrait, EntityTrait, IntoActiveModel, QueryFilter};
+use log::info;
+use redis_rate_limiter::redis::AsyncCommands;
+
+// TODO: Add some logic to check if the operating user is an admin
+// If he is, return true
+// If he is not, return false
+// This function is used to give permission to certain users
+
+
+pub async fn query_admin_modify_usertier<'a>(
+    app: &'a Web3ProxyApp,
+    bearer: Option<TypedHeader<Authorization<Bearer>>>,
+    params: &'a HashMap<String, String>
+) -> Result<Response, FrontendErrorResponse> {
+
+    // Quickly return if any of the input tokens are bad
+    let user_address: Vec<u8> = params
+        .get("user_address")
+        .ok_or_else(||
+            FrontendErrorResponse::StatusCode(
+                StatusCode::BAD_REQUEST,
+                "Unable to find user_address key in request".to_string(),
+                None,
+            )
+        )?
+        .parse::<Address>()
+        .map_err(|err| {
+            FrontendErrorResponse::StatusCode(
+                StatusCode::BAD_REQUEST,
+                "Unable to parse user_address as an Address".to_string(),
+                Some(err.into()),
+            )
+        })?
+        .to_fixed_bytes().into();
+    let user_tier_title = params
+        .get("user_tier_title")
+        .ok_or_else(|| FrontendErrorResponse::StatusCode(
+            StatusCode::BAD_REQUEST,
+            "Unable to get the user_tier_title key from the request".to_string(),
+            None,
+        ))?;
+
+    // Prepare output body
+    let mut response_body = HashMap::new();
+
+    // Establish connections
+    let db_conn = app.db_conn().context("query_admin_modify_user needs a db")?;
+    let db_replica = app
+        .db_replica()
+        .context("query_user_stats needs a db replica")?;
+    let mut redis_conn = app
+        .redis_conn()
+        .await
+        .context("query_admin_modify_user had a redis connection error")?
+        .context("query_admin_modify_user needs a redis")?;
+
+    // Try to get the user who is calling from redis (if existent) / else from the database
+    // TODO: Make a single query, where you retrieve the user, and directly from it the secondary user (otherwise we do two jumpy, which is unnecessary)
+    // get the user id first. if it is 0, we should use a cache on the app
+    let caller_id = get_user_id_from_params(&mut redis_conn, &db_conn, &db_replica, bearer, &params).await?;
+
+    // Check if the caller is an admin (i.e. if he is in an admin table)
+    let admin: admin::Model = admin::Entity::find()
+        .filter(admin::Column::UserId.eq(caller_id))
+        .one(&db_conn)
+        .await?
+        .context("This user is not registered as an admin")?;
```

Comment:
I think there's an AccessDenied frontend error that we should use. Using "context" is quick and easy, but it ends up being served as a 500. And 500s get automatically retried multiple times across the autoscaling group.

Reply:
Ok I think I added it in all the right places again. just lmk if you see a spot where it's off (i.e. supposed to be a 400 instead of a 500 or so)